Add an option to sort commits by author date in the revision grid

asfarley opened the pull request:
<!-- Please read CONTRIBUTING.md before submitting a pull request -->

Fixes #6826

Replace #6854


## Proposed changes
Add option to sort commits by author date in the revision grid. Add menu-option and Settings checkbox to access setting. Added tool-tip in Settings to warn of potential performance impact.


## Screenshots <!-- Remove this section if PR does not change UI -->

### Before
<img src="https://camo.githubusercontent.com/3c0443d6c994764bf0e0c34c8bc38b9baa904592/68747470733a2f2f692e696d6775722e636f6d2f486d6752446b6d2e706e67"/>

### After
<img src="https://camo.githubusercontent.com/c95ab46d3818b2dd69303858f253601655b8408b/68747470733a2f2f692e696d6775722e636f6d2f356f39536743532e706e67"/>


## Test methodology <!-- How did you ensure quality? -->
Tested on repo where commit-dates were not the same as author-dates. Appearance was as-expected. Tested enabling and disabling option; the commit sorting responded as expected.


## Test environment(s) <!-- Remove any that don't apply -->

GIT 2.17.1.windows.2
Windows NT 10.0.17763.0
.NET 4.7.3416.0

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

asfarley commented:
I've submitted a pull-request to GitExtensions documentation repo associated with this commit:
https://github.com/gitextensions/GitExtensionsDoc/pull/99

gerhardol commented:
Some comments from previous commit
https://github.com/gitextensions/gitextensions/pull/6854#issuecomment-514822536
The most important is to rebase on gitextensions/master

gerhardol commented on GitUI/CommandsDialogs/SettingsDialog/Pages/AppearanceSettingsPage.cs:
```diff
@@ -16,6 +16,7 @@ public partial class AppearanceSettingsPage : SettingsPageWithHeader
         private readonly TranslationString _noDictFile = new TranslationString("None");
         private readonly TranslationString _noDictFilesFound = new TranslationString("No dictionary files found in: {0}");
         private readonly TranslationString _noImageServiceTooltip = new TranslationString("A default image, if an email address has no matching Gravatar image.\r\nSee http://en.gravatar.com/site/implement/images#default-image for more details.");
+        private readonly TranslationString _authorDateSortWarningTooltip = new TranslationString("Sorting by author date may negatively impact performance.");
```
Suggestions:
Sorting by author date may delay displaying the revision graph

General performance is not affected

RussKie commented on GitUI/CommandsDialogs/SettingsDialog/Pages/AppearanceSettingsPage.cs:
```diff
@@ -16,6 +16,7 @@ public partial class AppearanceSettingsPage : SettingsPageWithHeader
         private readonly TranslationString _noDictFile = new TranslationString("None");
         private readonly TranslationString _noDictFilesFound = new TranslationString("No dictionary files found in: {0}");
         private readonly TranslationString _noImageServiceTooltip = new TranslationString("A default image, if an email address has no matching Gravatar image.\r\nSee http://en.gravatar.com/site/implement/images#default-image for more details.");
+        private readonly TranslationString _authorDateSortWarningTooltip = new TranslationString("Sorting by author date may negatively impact performance.");
```
```suggestion
        private readonly TranslationString _authorDateSortWarningTooltip = new TranslationString("Sorting by author date may delay rendering of the revision graph.");
```

gerhardol replied:
Agree that rendering is better

asfarley replied:
Fixed.

RussKie commented on GitUI/CommandsDialogs/SettingsDialog/Pages/AppearanceSettingsPage.Designer.cs:
```diff
@@ -120,14 +122,24 @@ private void InitializeComponent()
+            this.chkSortByAuthorDate.Location = new System.Drawing.Point(3, 99);
+            this.chkSortByAuthorDate.Name = "chkSortByAuthorDate";
+            this.chkSortByAuthorDate.Size = new System.Drawing.Size(116, 17);
+            this.chkSortByAuthorDate.TabIndex = 12;
```
Can you please confirm this is the logic tab order

asfarley replied:
I just tested this; the tab-order is correct.

asfarley commented:
Settings screenshot:

![Settings menu](https://user-images.githubusercontent.com/1622578/62230101-54b65f00-b37e-11e9-9db4-607756ed39f0.png)

asfarley commented:
Ok, I think I covered everything:

 * Rebased on the latest master
 * Fixed wording to clarify that rendering of revision grid is impacted, not general performance
 * Tested tab order

RussKie commented:
Thank you.
Can you please update https://github.com/gitextensions/GitExtensionsDoc/pull/99 to match.

RussKie commented:
@asfarley it looks like you haven't signed the contributors.txt.
Please sign it ASAP, otherwise we will revert your commit.

Thanks

gerhardol commented:
contributors now signed and merged
Thanks @asfarley
